src/borg/_hashindex.c

```diff
@@ -110,18 +110,35 @@ hashindex_index(HashIndex *index, const void *key)
     return _le32toh(*((uint32_t *)key)) % index->num_buckets;
 }
 
+inline int
+distance(int num_buckets, int current_idx, int ideal_idx)
+{
+    /* If the current index is smaller than the ideal index we've wrapped
+       around the end of the bucket array and need to compensate for that. */
+    return current_idx - ideal_idx + ( (current_idx < ideal_idx) ? num_buckets : 0 );
+}
+
+
 static int
-hashindex_lookup(HashIndex *index, const void *key)
+hashindex_lookup(HashIndex *index, const void *key, int *skip_hint)
 {
     int didx = -1;  // deleted index
     int start = hashindex_index(index, key);
     int idx = start;
-    for(;;) {
+    int num_buckets = index->num_buckets;
+    for(int offset = 0;; offset++) {
+        if (skip_hint != NULL) {
+            (*skip_hint) = offset;
+        }
         if(BUCKET_IS_EMPTY(index, idx))
         {
             return -1;
         }
+        if(offset > distance(num_buckets, idx, start)) {
+            return -1;
```

Comment:
@tw Will add. Basically this was an attempt to optimize lookup by taking advantage of the knowledge that the index is in RH order. So if the element you're looking for has larger offset-from-ideal-location than the bucket you're scanning now, you can give up.

This is the breaking change I mentioned.